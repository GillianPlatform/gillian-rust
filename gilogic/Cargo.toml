[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }