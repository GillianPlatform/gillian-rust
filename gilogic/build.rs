fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
