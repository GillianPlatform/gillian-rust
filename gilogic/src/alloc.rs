use std::alloc::{AllocError, Allocator, Layout};
use std::ptr::NonNull;

pub struct GillianAllocator;

#[gillian::no_translate]
#[gillian::builtin]
#[rustc_diagnostic_item = "gillian::alloc::alloc_array"]
pub fn alloc_array<T>(_len: usize) -> *mut T {
    unreachable!()
}

#[gillian::no_translate]
#[gillian::builtin]
#[rustc_diagnostic_item = "gillian::alloc::dealloc_array"]
pub fn dealloc_array<T>(_ptr: *mut T, _len: usize) -> *mut T {
    unreachable!()
}

#[allow(unused_variables)]
unsafe impl Allocator for GillianAllocator {
    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
        // Implemented in GIL
        panic!()
    }

    unsafe fn deallocate(&self, ptr: std::ptr::NonNull<u8>, layout: Layout) {
        // Implemented in GIL
        panic!()
    }

    fn allocate_zeroed(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
        // Implemented in GIL
        panic!()
    }
}
