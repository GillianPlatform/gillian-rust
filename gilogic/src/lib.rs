#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
