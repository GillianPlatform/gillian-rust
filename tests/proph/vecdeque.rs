//@check-pass
//?gil:ignore
#![feature(ptr_internals)]
#![feature(sized_type_properties)]
#![feature(allocator_api)]
#![feature(unchecked_math)]
extern crate gilogic;

use gilogic::{
    __stubs::{PointsToMaybeUninit, PointsToSlice},
    alloc::GillianAllocator,
    iterated::{all_own, all_own_swap},
    macros::{
        assertion, extract_lemma, predicate, prophecies::with_freeze_lemma_for_mutref,
        specification,
    },
    mutref_auto_resolve,
    prophecies::{Ownable, Prophecised, Prophecy},
    Seq,
};
use std::alloc::{Allocator, Layout, LayoutError};
use std::mem::{self, SizedTypeProperties};
use std::ptr::{NonNull, Unique};
use std::slice;

macro_rules! branch {
    ($cond:expr) => {
        if $cond {
        } else {
        }
    };
}

pub struct TryReserveError {
    kind: TryReserveErrorKind,
}

impl From<TryReserveErrorKind> for TryReserveError {
    fn from(kind: TryReserveErrorKind) -> Self {
        Self { kind }
    }
}

impl TryReserveError {
    fn kind(&self) -> TryReserveErrorKind {
        self.kind
    }
}

#[derive(Clone, Copy)]
pub enum TryReserveErrorKind {
    /// Error due to the comxputed capacity exceeding the collection's maximum
    /// (usually `isize::MAX` bytes).
    CapacityOverflow,
    AllocError {
        layout: Layout,
        non_exhaustive: (),
    },
}

use TryReserveErrorKind::*;

enum AllocInit {
    /// The contents of the new memory are uninitialized.
    Uninitialized,
    /// The new memory is guaranteed to be zeroed.
    Zeroed,
}

impl Ownable for AllocInit {
    type RepresentationTy = Self;

    #[predicate]
    fn own(self, model: Self::RepresentationTy) {
        assertion!((self == model))
    }
}

fn handle_reserve(result: Result<(), TryReserveError>) {
    match result {
        Err(err) => match err.kind() {
            CapacityOverflow => capacity_overflow(),
            AllocError { layout, .. } => handle_alloc_error(layout),
        },
        Ok(()) => { /* yay */ }
    }
}

// #[lemma]
// #[requires(|content: Seq<Option<T>>, repr: Seq<Option<T::RepresentationTy>>|
//             (before < after) * ptr.many_maybe_uninit(before, content) * all_own(content, repr) *
//             ptr.add(before).many_uninits(after - before))]
// #[ensures(|nones: Seq<Option<T>>, nones_r: Seq<Option<T::RepresentationTy>>|
//             ptr.many_maybe_uninit(after, content.concat(nones)) * all_own(content.concat(nones), repr.concat(nones_r)) *
//             (nones.len() == after - before) * (nones.len() == nones_r.len()) *
//             all_none(nones) * all_none(nones_r))]
// fn concat_own_uninit<T: Ownable>(ptr: *mut T, before: usize, after: usize);

// #[lemma]
// #[requires(ptr.many_uninits(cap))]
// #[ensures(
//         ptr.many_maybe_uninit(cap, Seq::<Option<T>>::repeat(None, cap)) *
//         all_own(Seq::<Option<T>>::repeat(None, cap), Seq::<Option<T::RepresentationTy>>::repeat(None, cap))
//     )]
// fn uninit_to_raw_vec<T: Ownable>(ptr: *mut T, cap: usize);

pub(crate) struct RawVec<T> {
    ptr: Unique<T>,
    cap: usize,
}

fn finish_grow(
    new_layout: Result<Layout, LayoutError>,
    current_memory: Option<(NonNull<u8>, Layout)>,
) -> Result<NonNull<[u8]>, TryReserveError> {
    // Check for the error here to minimize the size of `RawVec::grow_*`.
    let new_layout = match new_layout {
        Ok(x) => x,
        Err(_) => {
            return Err(TryReserveError {
                kind: CapacityOverflow,
            })
        }
    };

    match alloc_guard(new_layout.size()) {
        Ok(x) => x,
        Err(e) => return Err(e),
    };

    let memory = if let Some((ptr, old_layout)) = current_memory {
        // debug_assert_eq!(old_layout.align(), new_layout.align());
        unsafe {
            // The allocator checks for alignment equality
            assert!(old_layout.align() == new_layout.align());
            GillianAllocator.grow(ptr, old_layout, new_layout)
        }
    } else {
        GillianAllocator.allocate(new_layout)
    };

    match memory {
        Ok(x) => Ok(x),
        Err(_) => Err(AllocError {
            layout: new_layout,
            non_exhaustive: (),
        }
        .into()),
    }
}

impl<T> RawVec<T> {
    const MIN_NON_ZERO_CAP: usize = if mem::size_of::<T>() == 1 {
        8
    } else if mem::size_of::<T>() <= 1024 {
        4
    } else {
        1
    };

    pub const fn new() -> Self {
        Self {
            ptr: Unique::dangling(),
            cap: 0,
        }
    }

    pub const NEW: Self = Self::new();

    fn current_memory(&self) -> Option<(NonNull<u8>, Layout)> {
        if T::IS_ZST || self.cap == 0 {
            None
        } else {
            unsafe {
                let align = 1; // We don't support align yet, but plan on supporting them exactly in the same way we support size_of.
                               // It's just more implementation and lack of type, but the skeleton is already there in LayoutKnowlege.
                let size = mem::size_of::<T>().unchecked_mul(self.cap);
                let layout = Layout::from_size_align_unchecked(size, align);
                Some((self.ptr.cast().into(), layout))
            }
        }
    }

    // #[requires(init.own(AllocInit::Uninitialized) * capacity.own(capacity) * (capacity < 2 << 62))]
    // #[ensures(ret.own(Seq::repeat(None, capacity)))]
    fn allocate(capacity: usize, init: AllocInit) -> Self {
        // Don't allocate here because `Drop` will not deallocate when `capacity` is 0.
        if T::IS_ZST || capacity == 0 {
            Self::new()
        } else {
            let layout = match Layout::array::<T>(capacity) {
                Ok(layout) => layout,
                Err(_) => capacity_overflow(),
            };
            match alloc_guard(layout.size()) {
                Ok(_) => {}
                Err(_) => capacity_overflow(),
            }
            let result = match init {
                AllocInit::Uninitialized => GillianAllocator.allocate(layout),
                AllocInit::Zeroed => GillianAllocator.allocate_zeroed(layout),
            };
            let ptr = match result {
                Ok(ptr) => ptr,
                Err(_) => handle_alloc_error(layout),
            };
            let ptr = unsafe { Unique::new_unchecked(ptr.cast().as_ptr()) };
            Self { ptr, cap: capacity }
        }
    }

    pub fn ptr(&self) -> *mut T {
        self.ptr.as_ptr()
    }

    // #[requires(|current: <Self as Ownable>::RepresentationTy, future: <Self as Ownable>::RepresentationTy| self.own((current, future)) * len.own(len) * additional.own(additional) * (additional > 0))]
    // #[ensures(
    //         $future.len() - current.len() >= additional$ // Note quite that
    //         $future == current.concat(Seq::repeat(future.len() - current.len()))$
    //  )]
    fn grow_amortized(&mut self, len: usize, additional: usize) -> Result<(), TryReserveError> {
        // This is ensured by the calling contexts.
        // assert!(additional > 0);

        if T::IS_ZST {
            // Since we return a capacity of `usize::MAX` when `elem_size` is
            // 0, getting to here necessarily means the `RawVec` is overfull.
            return Err(TryReserveError {
                kind: CapacityOverflow,
            });
        }

        // Nothing we can really do about these checks, sadly.
        let required_cap = match len.checked_add(additional) {
            Some(x) => x,
            None => {
                return Err(TryReserveError {
                    kind: CapacityOverflow,
                })
            }
        };

        // This guarantees exponential growth. The doubling cannot overflow
        // because `cap <= isize::MAX` and the type of `cap` is `usize`.
        let cap = std::cmp::max(self.cap * 2, required_cap);
        let cap = std::cmp::max(Self::MIN_NON_ZERO_CAP, cap);

        let new_layout = Layout::array::<T>(cap);

        // `finish_grow` is non-generic over `T`.
        let ptr = match finish_grow(new_layout, self.current_memory()) {
            Ok(ptr) => ptr,
            Err(err) => return Err(err),
        };
        self.ptr = unsafe { Unique::new_unchecked(ptr.cast().as_ptr()) };
        self.cap = cap;
        Ok(())
    }

    pub fn capacity(&self) -> usize {
        if T::IS_ZST {
            usize::MAX
        } else {
            self.cap
        }
    }

    pub fn reserve_for_push(&mut self, len: usize) {
        handle_reserve(self.grow_amortized(len, 1));
    }

    // #[requires(capacity.own(capacity) * (capacity < 2 << 62))]
    // #[ensures(|model| ret.own(model) * (model == Seq::repeat(None, model.len())))]
    pub fn with_capacity(capacity: usize) -> Self {
        Self::allocate(capacity, AllocInit::Uninitialized)
    }
}

fn capacity_overflow() -> ! {
    // Capacity overflow!
    panic!();
}

#[allow(unused_variables)]
fn handle_alloc_error(layout: Layout) -> ! {
    // Allocation failed!
    panic!()
}

fn alloc_guard(alloc_size: usize) -> Result<(), TryReserveError> {
    if usize::BITS < 64 && alloc_size > isize::MAX as usize {
        Err(TryReserveErrorKind::CapacityOverflow.into())
    } else {
        Ok(())
    }
}

pub struct VecDeque<T> {
    buf: RawVec<T>,
    len: usize,
    head: usize,
}

// #[extract_lemma(
//     forall ptr: Unique<T>, cap, len.
//     model m.
//     extract model mh: (T::RepresentationTy, T::RepresentationTy).
//     assuming { ix < len }
//     from { vec_ref_mut_pcl(vec, m, ptr, cap, len) }
//     extract { Ownable::own(&mut (*(ptr.as_ptr().add(ix))), mh) }
//     prophecise { m.sub(0, ix).append(mh).concat(m.sub(ix + 1, len - ix - 1)) }
// )]
// fn extract_ith<'a, T: Ownable>(vec: &'a mut Vec<T>, ix: usize) -> Prophecy<T::RepresentationTy>;

// #[with_freeze_lemma_for_mutref(
//     lemma_name = freeze_pcl,
//     predicate_name = vec_ref_mut_pcl,
//     frozen_variables = [ptr, cap, len],
//     inner_predicate_name = vec_ref_mut_inner_pcl,
//     resolve_macro_name = auto_resolve_vec_ref_mut_pcl
// )]
impl<T: Ownable> Ownable for VecDeque<T> {
    type RepresentationTy = Seq<T::RepresentationTy>;

    #[predicate]
    fn own(self, model: Self::RepresentationTy) {
        assertion!(
            |ptr: Unique<T>, cap: usize, len: usize, head: usize, values, rest, rest2| (head + len
                <= cap)
                * (std::mem::size_of::<T>() > 0)
                * (self
                    == VecDeque {
                        buf: RawVec { ptr, cap },
                        len,
                        head
                    })
                * cap.own(cap)
                * len.own(len)
                * head.own(head)
                * ptr.as_ptr().add(head).points_to_slice(len, values)
                * (len == values.len())
                * (values.len() == model.len())
                * all_own(values, model)
                * ptr
                    .as_ptr()
                    .add(head + len)
                    .many_maybe_uninits(cap - len, rest)
                * ptr.as_ptr().many_maybe_uninits(head, rest2)
        );
        //                                      head                     cap
        //      +------------------+-------------+------------------------+
        //      | head + len - cap |             |   cap - head           |
        //      +------------------+-------------+------------------------+
        assertion!(|ptr: Unique<T>,
                    cap: usize,
                    len: usize,
                    head: usize,
                    values,
                    values_head: Seq<_>,
                    values_tail,
                    end,
                    rest| (head + len > cap)
            * (head < cap)
            * (len <= cap)
            * (std::mem::size_of::<T>() > 0)
            * (self
                == VecDeque {
                    buf: RawVec { ptr, cap },
                    len,
                    head
                })
            * (end == (head + len - cap))
            * (values == values_head.concat(values_tail))
            * cap.own(cap)
            * len.own(len)
            * head.own(head)
            * ptr
                .as_ptr()
                .add(head)
                .points_to_slice(cap - head, values_head)
            * ptr.as_ptr().points_to_slice(end, values_tail)
            * (len == values.len())
            * (values.len() == model.len())
            * all_own(values, model)
            * ptr.as_ptr().add(end).many_maybe_uninits(cap - len, rest));
    }
}

impl<T: Ownable> VecDeque<T> {
    #[specification(
        requires { emp }
        ensures { ret.own(Seq::empty())}
    )]
    pub const fn new() -> Self {
        branch!(T::IS_ZST);
        VecDeque {
            buf: RawVec::NEW,
            head: 0,
            len: 0,
        }
    }

    #[specification(
        requires { capacity.own(capacity) * (capacity <= isize::MAX as usize)}
        ensures { ret.own(Seq::empty()) }
    )]
    pub fn with_capacity(capacity: usize) -> Self {
        VecDeque {
            buf: RawVec::with_capacity(capacity),
            head: 0,
            len: 0,
        }
    }

    pub fn len(&self) -> usize {
        self.len
    }

    pub fn capacity(&self) -> usize {
        self.buf.capacity()
    }

    #[inline]
    fn is_full(&self) -> bool {
        self.len == self.capacity()
    }

    #[creusillian::requires((*self)@.len() < 1000 )]
    #[creusillian::ensures((^self)@ == (*self)@.append(value))]
    pub fn push_back(&mut self, value: T) {
        if self.is_full() {
            self.grow();
        }

        unsafe { self.buffer_write(self.to_physical_idx(self.len()), value) }
        self.len += 1;
        mutref_auto_resolve!(self);
    }

    fn grow(&mut self) {
        // debug_assert!(self.is_full());
        let old_cap = self.capacity();

        self.buf.grow_amortized(self.len, 1);
        unsafe {
            self.handle_capacity_increase(old_cap);
        }

        // debug_assert!(!self.is_full());
    }

    /// Writes an element into the buffer, moving it.
    #[inline]
    unsafe fn buffer_write(&mut self, off: usize, value: T) {
        unsafe {
            // std::ptr::write(self.ptr().add(off), value);
        }
    }

    #[inline]
    fn to_physical_idx(&self, idx: usize) -> usize {
        self.wrap_add(self.head, idx)
    }

    #[inline]
    fn wrap_add(&self, idx: usize, addend: usize) -> usize {
        wrap_index(idx.wrapping_add(addend), self.capacity())
    }

    /// Frobs the head and tail sections around to handle the fact that we
    /// just reallocated. Unsafe because it trusts old_capacity.
    #[inline]
    unsafe fn handle_capacity_increase(&mut self, old_capacity: usize) {
        let new_capacity = self.capacity();
        // debug_assert!(new_capacity >= old_capacity);

        // Move the shortest contiguous section of the ring buffer
        //
        // H := head
        // L := last element (`self.to_physical_idx(self.len - 1)`)
        //
        //    H             L
        //   [o o o o o o o o ]
        //    H             L
        // A [o o o o o o o o . . . . . . . . ]
        //        L H
        //   [o o o o o o o o ]
        //          H             L
        // B [. . . o o o o o o o o . . . . . ]
        //              L H
        //   [o o o o o o o o ]
        //              L                 H
        // C [o o o o o o . . . . . . . . o o ]

        // can't use is_contiguous() because the capacity is already updated.
        if self.head <= old_capacity - self.len {
            // A
            // Nop
        } else {
            let head_len = old_capacity - self.head;
            let tail_len = self.len - head_len;
            if head_len > tail_len && new_capacity - old_capacity >= tail_len {
                // B
                unsafe {
                    self.copy_nonoverlapping(0, old_capacity, tail_len);
                }
            } else {
                // C
                let new_head = new_capacity - head_len;
                unsafe {
                    // can't use copy_nonoverlapping here, because if e.g. head_len = 2
                    // and new_capacity = old_capacity + 1, then the heads overlap.
                    self.copy(self.head, new_head, head_len);
                }
                self.head = new_head;
            }
        }
        // debug_assert!(self.head < self.capacity() || self.capacity() == 0);
    }

    #[inline]
    unsafe fn copy_nonoverlapping(&mut self, src: usize, dst: usize, len: usize) {
        unsafe {
            std::ptr::copy_nonoverlapping(self.ptr().add(src), self.ptr().add(dst), len);
        }
    }

    unsafe fn copy(&mut self, src: usize, dst: usize, len: usize) {
        unsafe {
            std::ptr::copy(self.ptr().add(src), self.ptr().add(dst), len);
        }
    }

    #[inline]
    fn ptr(&self) -> *mut T {
        self.buf.ptr()
    }
}

#[inline]
fn wrap_index(logical_index: usize, capacity: usize) -> usize {
    // debug_assert!(
    //     (logical_index == 0 && capacity == 0)
    //         || logical_index < capacity
    //         || (logical_index - capacity) < capacity
    // );
    if logical_index >= capacity {
        logical_index - capacity
    } else {
        logical_index
    }
}
