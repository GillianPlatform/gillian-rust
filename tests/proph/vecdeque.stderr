warning: unused imports: `Prophecised`, `Prophecy`
  --> ../tests/proph/vecdeque.rs:15:27
   |
15 |     prophecies::{Ownable, Prophecised, Prophecy},
   |                           ^^^^^^^^^^^  ^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `std::slice`
  --> ../tests/proph/vecdeque.rs:21:5
   |
21 | use std::slice;
   |     ^^^^^^^^^^

warning: unnecessary parentheses around function argument
   --> ../tests/proph/vecdeque.rs:369:23
    |
369 |             * (end == (head + len - cap))
    |                       ^                ^
    |
    = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
    |
369 -             * (end == (head + len - cap))
369 +             * (end == head + len - cap)
    |

[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"std"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"creusillian"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _2
warning: unused variable: `off`
   --> ../tests/proph/vecdeque.rs:451:39
    |
451 |     unsafe fn buffer_write(&mut self, off: usize, value: T) {
    |                                       ^^^ help: if this is intentional, prefix it with an underscore: `_off`
    |
    = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `value`
   --> ../tests/proph/vecdeque.rs:451:51
    |
451 |     unsafe fn buffer_write(&mut self, off: usize, value: T) {
    |                                                   ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`

[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _3
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _11
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _2
warning: the feature `unchecked_math` has been stable since 1.79.0 and no longer requires an attribute to enable
 --> ../tests/proph/vecdeque.rs:6:12
  |
6 | #![feature(unchecked_math)]
  |            ^^^^^^^^^^^^^^
  |
  = note: `#[warn(stable_features)]` on by default

warning: unnecessary `unsafe` block
   --> ../tests/proph/vecdeque.rs:452:9
    |
452 |         unsafe {
    |         ^^^^^^ unnecessary `unsafe` block
    |
    = note: `#[warn(unused_unsafe)]` on by default

warning: field `kind` is never read
  --> ../tests/proph/vecdeque.rs:32:5
   |
31 | pub struct TryReserveError {
   |            --------------- field in this struct
32 |     kind: TryReserveErrorKind,
   |     ^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: method `kind` is never used
  --> ../tests/proph/vecdeque.rs:42:8
   |
41 | impl TryReserveError {
   | -------------------- method in this implementation
42 |     fn kind(&self) -> TryReserveErrorKind {
   |        ^^^^

warning: variant `Zeroed` is never constructed
  --> ../tests/proph/vecdeque.rs:64:5
   |
60 | enum AllocInit {
   |      --------- variant in this enum
...
64 |     Zeroed,
   |     ^^^^^^

warning: function `handle_reserve` is never used
  --> ../tests/proph/vecdeque.rs:76:4
   |
76 | fn handle_reserve(result: Result<(), TryReserveError>) {
   |    ^^^^^^^^^^^^^^

warning: method `reserve_for_push` is never used
   --> ../tests/proph/vecdeque.rs:265:12
    |
149 | impl<T> RawVec<T> {
    | ----------------- method in this implementation
...
265 |     pub fn reserve_for_push(&mut self, len: usize) {
    |            ^^^^^^^^^^^^^^^^

warning: unused `std::result::Result` that must be used
   --> ../tests/proph/vecdeque.rs:441:9
    |
441 |         self.buf.grow_amortized(self.len, 1);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this `Result` may be an `Err` variant, which should be handled
    = note: `#[warn(unused_must_use)]` on by default
help: use `let _ = ...` to ignore the resulting value
    |
441 |         let _ = self.buf.grow_amortized(self.len, 1);
    |         +++++++

warning: 13 warnings emitted

[TIMESTAMP DEBUG rust_to_gil] Correct!
